src/hooks/cgroups/pbs_cgroups.PY

```diff
@@ -3089,35 +3090,30 @@ class CgroupUtils(object):
                         empty_cores.add(corenum)
                 # Assign threads from the empty cores
                 for corenum in empty_cores:
+                    assigned['cpuset.cpus'].append(corenum)
```

Comment:
I don't fully understand how the data is stored, but if we have a node with 4 empty cores, each with 2 threads, we should have 8 threads. If we request 4 cpus, then needed = 4.
We check the first core, append it to `cpuset.cpus`, remove it from `cores`, then reduce `needed` by 1 `\\needed = 3`. 
We check the first thread of first core, append it to `cpuset.cpus`, remove it from `cores`, then reduce `needed` by 1 `\\needed = 2`. 
We check the second thread of first core, append it to `cpuset.cpus`, remove it from `cores`, then reduce `needed` by 1 `\\needed = 1`. 
We check the second core, append it to `cpuset.cpus`, remove it from `cores`, then reduce `needed` by 1 `\\needed = 0`. Because `needed == 0`, we break out. 

At the end, we have core1, core1threadA, core1threadB, and core2. 
Shouldn't it be core1threadA, core1threadB, core2threadA, core2threadB?

Reply:
The _discover_cpuinfo() method is responsible for populating self.cpuinfo in the NodeConfig class. The available_node_resources() method then removes assigned resources from the dictionary, resulting in the "available" dictionary that gets passed to _assign_resources(). At this point each physical CPU has a thread list associated with it (node.cpuinfo['cpu'][corenum]['threads']). When use_hyperthreads is enabled, the hook attempts to assign empty core/thread sets before it resorts to assigning those that are partially allocated.

Using your example of an empty 4 core (8 thread) node...
- All four cores are identified as empty
- The first core is assigned to the job (needed = 3)
- The thread associated with the previously assigned core is assigned to the job (needed = 2)
- The second core is assigned to the job (needed = 1)
- The thread associated with the second core is assigned to the job (needed = 0)
- We break from the loop because needed == 0

Ultimately, we end up with what you thought we should... core1threadA, core1threadB, core2threadA, core2threadB.

The test was failing on a hyperthread enabled system because it was attempting to place two jobs on different threads of the same core. The hook used to (erroneously) reject the request. It was a corner case when ncpus=1 and there was only one thread available to assign.

Reply:
I see; so each cpu has a thread, not two threads.

Reply:
Correct.